vscode/src/jsonrpc/agent-protocol.ts

```diff
@@ -246,7 +246,7 @@ export type ServerRequests = {
     'window/showMessage': [ShowWindowMessageParams, string | null]
 
     'textDocument/edit': [TextDocumentEditParams, boolean]
-    'textDocument/openUntitledDocument': [UntitledTextDocument, boolean]
+    'textDocument/openDocument': [{ uri: string }, ProtocolTextDocument]
```

Comment:
I lean towards not returning the actual document here. Instead, the contract of this requests should be: the client returns true if it sent a `textDocument/didOpen` notification, and false otherwise. The server should read the document properties from `textDocument/didOpen` instead of the return value here

```suggestion
    'textDocument/openDocument': [{ uri: string }, boolean]
```

Reply:
The issue is that we then in agent have code paths that don't have document data yet right after opening. That was causing the desync

Reply:
The contract is that the client must have sent the notification before returning true. To implement `vscode.workspace.openTextDocument`, you need to await on the `true` response and then read from `AgentWorkspaceDocument`.

Reply:
What would be the reason to implement it that way?
VSCode contract as I read it does not really guarantee you that `didOpen` will shoot before you return:
`Opens a document. Will return early if this document is already open. Otherwise the document is loaded and the didOpen-event fires.`
It just guarantees it will fire.

I think making signature of this method similar to:
`export function openTextDocument(uri: Uri): Thenable<TextDocument>;`
makes sense.
I'm open to discussion, but I'm not sure I see the benefits of the proposed change Olaf.

Reply:
The VSC `didOpen` notification has a different contract from agent-protocol `didOpen`. 

> I'm open to discussion, but I'm not sure I see the benefits of the proposed change Olaf.

The motivation to have a single source of truth on document synchronization. Allowing `openDocument` to return text contents is an invitation to deal with desync issues. It's easier to reason about the system if all document synchronization happens via `textDocument/{didOpen,didChange,didClose,didFocus}` notifications.

Reply:
Changes as we discussed.